04-Parallel-Simulations-With-MPI/02-Domain-Decomposition.ipynb

```diff
@@ -0,0 +1,446 @@
+       "\u001b[0;34m\u001b[0m\u001b[0mintegrator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmethods\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnvt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0msim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moperations\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mintegrator\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mintegrator\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0msim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0mtps\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msim\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtps\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m\u001b[0;32mif\u001b[0m \u001b[0mdevice\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrank\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\n",
+       "\u001b[0;34m\u001b[0m    \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtps\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n"
+      ]
+     },
+     "metadata": {},
+     "output_type": "display_data"
+    }
+   ],
+   "source": [
+    "%pycat lj_performance2.py"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "notice(2): Using domain decomposition: n_x = 1 n_y = 2 n_z = 2.\n",
+      "2496.398944522526\n"
+     ]
+    }
+   ],
+   "source": [
+    "!mpirun -n 4 python3 lj_performance2.py"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note the pattern used here:\n",
+    "```python\n",
+    "tps = sim.tps\n",
+    "if device.communicator.rank == 0:\n",
+    "    print(tps)\n",
+    "```\n",
+    "The property `sim.tps` is accessed on all ranks, but printed only on rank 0.\n",
+    "\n",
+    "If the code was instead like that below the property `sim.tps` would be accessed only on rank 0 and would **deadlock** the execution.:\n",
+    "```python\n",
+    "if device.communicator.rank == 0:\n",
+    "    print(sim.tps)\n",
+    "```\n",
+    "\n",
+    "So, be careful using `if device.communicator.rank == 0:`.\n",
```

Comment:
I'm still not sure why this would result in deadlock. I can image that if I do `sim.run(1000)` and then `if device.communicator.rank == 0: sim.run(1000)`, this would mess up the communication because all other ranks are at sweep 1000 but rank 0 is at sweep 2000. Could you explain a little more why all other ranks would have to wait in this example

Reply:
TPS is a poor example here because it technically doesn't result in a deadlock. I will replace this with the example of printing the current system kinetic energy, which must be summed across all ranks.

Reply:
I also moved the more complex deadlock example later, after explaining domain decomposition since that topic does need to be explained first.